source/TestStep.cpp

```diff
@@ -0,0 +1,287 @@
+{
+	switch (stepType)
+	{
+		case TestStep::ASSERT:
+		case TestStep::WAITFOR:
+			// If we reached the condition, then done
+			if (checkedCondition.Test(player.Conditions()))
+				return RESULT_DONE;
+			if (stepType == TestStep::ASSERT)
+				return RESULT_FAIL;
+			// In this case, we are waiting for the condition
+			return RESULT_RETRY;
+			break;
+		case TestStep::LAUNCH:
+			// If flying around, then launching the ship succesfully happened
+			if (PlayerIsFlyingAround(menuPanels, gamePanels, player))
+				return RESULT_DONE;
+			// Should implement some function to close this menu. But
+			// fail for now if the player/game menu is active.
+			if (PlayerMenuIsActive(menuPanels))
+				return RESULT_FAIL;
+			if (PlayerOnPlanetMainScreen(menuPanels, gamePanels, player)){
+				// Launch using the conversation mechanism. Not the most
+				// appropriate way to launch, but works for now.
+				player.BasicCallback(Conversation::LAUNCH);
+				return RESULT_RETRY;
+			}
+			// No idea where we are and what we are doing. But we are not
+			// in a position to launch, so fail this teststep.
+			return RESULT_FAIL;
+			break;
+		case TestStep::LAND:
+			// If the player/game menu is active, then we are not in a state
+			// where land makes sense.
+			if (PlayerMenuIsActive(menuPanels))
+				return RESULT_FAIL;
+			// If we are still flying around, then we are not on a planet.
+			if (PlayerIsFlyingAround(menuPanels, gamePanels, player))
+			{
+				Ship * playerFlagShip = player.Flagship();
+				if (playerFlagShip == nullptr)
+					return RESULT_FAIL;
+				if (stepAction == 0)
+				{
+					// Send the land command here
+					// Player commands are handled in Engine.cpp (at the moment this code was updated)
+					// TODO: Landing should also have a stellar target (just to reduce ambiguity)
+					Command command;
+					command.Set(Command::LAND);
+					if(!SendFlightCommand(command, gamePanels))
+						return RESULT_FAIL;
+					stepAction++;
+				}
+				return RESULT_NEXTACTION;
+			}
+			if (PlayerOnPlanetMainScreen(menuPanels, gamePanels, player))
+				return RESULT_DONE;
+
+			// Unknown state/screen. Landing fails.
+			return RESULT_FAIL;
+			break;
+		case TestStep::LOAD_GAME:
+			if (stepAction == 0){
+				// Check if the savegame actually exists
+				if (! Files::Exists(Files::Saves() + FilePathOrName()))
+					return RESULT_FAIL;
+				// Perform the load and verify that player is loaded.
+				player.Load(Files::Saves() + FilePathOrName());
+				if (!player.IsLoaded())
+					return RESULT_FAIL;
+				// Enable testmode to prevent automatic saving.
+				player.SetTestMode();
+				// Actual load succeeded. Allow game to adopt to new
+				// situation and then continue with enter/pilot step.
+				return RESULT_NEXTACTION;
+			}
+			else if (stepAction == 1)
+			{
+				// Clear the menu entries and go to main game screen
+				if (! menuPanels.IsEmpty())
+					menuPanels.Pop(menuPanels.Top().get());
```

Comment:
We should send keystrokes / commands that the player actions perform where possible.

Reply:
Yes; I already tested #4715 in combination with this PR today to see that commands can be given from the TestRunner to the Engine/Fleet/flagShip using the same commands as that players would send.
A next step is to do the same for planetary actions (feature-request #4530 is related to that) and loading/saving or just generic keyboard input should also work this way where possible.

The in-flight and on-planet actions will take quite a bit of attention first, so I'll add a `//TODO` about improving this part of the code to make sure that I don't forget this.

Reply:
I handled this comment by adding a `//TODO` item for now.